ticketbuyer/purchase.go

```diff
@@ -65,4 +74,25 @@ out:
 			break out
 		}
 	}
+	p.wg.Done()
+}
+
+// Start starts the purchase manager goroutines.
+func (p *PurchaseManager) Start() {
+	p.wg.Add(1)
+	go p.NotificationHandler()
+}
+
+// WaitForShutdown blocks until all purchase manager goroutines have finished executing.
+func (p *PurchaseManager) WaitForShutdown() {
```

Comment:
Tried to solve this by adding a new struct `waiter`. It used to set the wait function to the `WaitForShutdown` function of the latest `PurchaseManager` instance.